// @dharmaprotocol/dharma.js/dist/lib/src/adapters/simple_interest_loan_adapter.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var singleLineString = require("single-line-string");
var omit = require("lodash.omit");
// utils
var bignumber_1 = require("../../utils/bignumber");
var constants_1 = require("../../utils/constants");
// types
var types_1 = require("../types");
var invariants_1 = require("../invariants");
var AmortizationUnitCodes = ["hours", "days", "weeks", "months", "years"];
var AmortizationUnitCode;
(function (AmortizationUnitCode) {
    AmortizationUnitCode[AmortizationUnitCode["HOURS"] = 0] = "HOURS";
    AmortizationUnitCode[AmortizationUnitCode["DAYS"] = 1] = "DAYS";
    AmortizationUnitCode[AmortizationUnitCode["WEEKS"] = 2] = "WEEKS";
    AmortizationUnitCode[AmortizationUnitCode["MONTHS"] = 3] = "MONTHS";
    AmortizationUnitCode[AmortizationUnitCode["YEARS"] = 4] = "YEARS";
})(AmortizationUnitCode || (AmortizationUnitCode = {}));
var MAX_PRINCIPAL_TOKEN_INDEX_HEX = "0xff";
var MAX_PRINCIPAL_AMOUNT_HEX = "0xffffffffffffffffffffffff";
var MAX_TERM_LENGTH_VALUE_HEX = "0xffff";
var MAX_INTEREST_RATE_PRECISION = 4;
var FIXED_POINT_SCALING_FACTOR = Math.pow(10, MAX_INTEREST_RATE_PRECISION);
var MAX_INTEREST_RATE = Math.pow(2, 24) / FIXED_POINT_SCALING_FACTOR;
exports.SimpleInterestAdapterErrors = {
    INVALID_TOKEN_INDEX: function (tokenIndex) {
        return singleLineString(templateObject_1 || (templateObject_1 = __makeTemplateObject(["Token Registry does not track a token at index\n                         ", "."], ["Token Registry does not track a token at index\n                         ", "."])), tokenIndex.toString());
    },
    INVALID_PRINCIPAL_AMOUNT: function () {
        return singleLineString(templateObject_2 || (templateObject_2 = __makeTemplateObject(["Principal amount must be a whole number greater than 0\n                         and less than 2^96 - 1."], ["Principal amount must be a whole number greater than 0\n                         and less than 2^96 - 1."])));
    },
    INVALID_INTEREST_RATE: function () {
        return singleLineString(templateObject_3 || (templateObject_3 = __makeTemplateObject(["Interest amount cannot be negative,\n                         greater than 1677.7216, or have more than\n                         ", " decimal places."], ["Interest amount cannot be negative,\n                         greater than 1677.7216, or have more than\n                         ", " decimal places."])), MAX_INTEREST_RATE_PRECISION);
    },
    INVALID_AMORTIZATION_UNIT_TYPE: function () {
        return singleLineString(templateObject_4 || (templateObject_4 = __makeTemplateObject(["Amortization unit must be of type HOURS, DAYS,\n                         WEEKS, MONTHS, or YEARS."], ["Amortization unit must be of type HOURS, DAYS,\n                         WEEKS, MONTHS, or YEARS."])));
    },
    INVALID_TERM_LENGTH: function () {
        return singleLineString(templateObject_5 || (templateObject_5 = __makeTemplateObject(["Term length value cannot be negative or greater\n                         than 2^120 - 1"], ["Term length value cannot be negative or greater\n                         than 2^120 - 1"])));
    },
    INVALID_TERMS_CONTRACT: function (principalToken, termsContract) {
        return singleLineString(templateObject_6 || (templateObject_6 = __makeTemplateObject(["Terms Contract at address ", " does not\n                         correspond to the SimpleInterestTermsContract associated\n                         with the principal token at address ", ""], ["Terms Contract at address ", " does not\n                         correspond to the SimpleInterestTermsContract associated\n                         with the principal token at address ", ""])), termsContract, principalToken);
    },
    UNSUPPORTED_PRINCIPAL_TOKEN: function (principalTokenSymbol) {
        return singleLineString(templateObject_7 || (templateObject_7 = __makeTemplateObject(["Token with symbol ", " is not supported\n                         by the Dharma Token Registry"], ["Token with symbol ", " is not supported\n                         by the Dharma Token Registry"])), principalTokenSymbol);
    },
    MISMATCHED_TOKEN_SYMBOL: function (principalTokenAddress, symbol) {
        return singleLineString(templateObject_8 || (templateObject_8 = __makeTemplateObject(["Terms contract parameters are invalid for the given debt order.\n                         Principal token at address ", " does not\n                         correspond to specified token with symbol ", ""], ["Terms contract parameters are invalid for the given debt order.\n                         Principal token at address ", " does not\n                         correspond to specified token with symbol ", ""])), principalTokenAddress, symbol);
    },
    MISMATCHED_TERMS_CONTRACT: function (termsContract) {
        return singleLineString(templateObject_9 || (templateObject_9 = __makeTemplateObject(["Terms contract at address ", " is not\n                         a SimpleInterestTermsContract.  As such, this adapter will not\n                         interface with the terms contract as expected"], ["Terms contract at address ", " is not\n                         a SimpleInterestTermsContract.  As such, this adapter will not\n                         interface with the terms contract as expected"])), termsContract);
    },
};
var TX_DEFAULTS = { from: constants_1.NULL_ADDRESS, gas: 0 };
var SimpleInterestLoanTerms = /** @class */ (function () {
    function SimpleInterestLoanTerms(web3, contracts) {
        this.assert = new invariants_1.Assertions(web3, contracts);
    }
    SimpleInterestLoanTerms.prototype.packParameters = function (termsContractParameters) {
        var principalTokenIndex = termsContractParameters.principalTokenIndex, principalAmount = termsContractParameters.principalAmount, interestRate = termsContractParameters.interestRate, amortizationUnit = termsContractParameters.amortizationUnit, termLength = termsContractParameters.termLength;
        this.assertPrincipalTokenIndexWithinBounds(principalTokenIndex);
        this.assertPrincipalAmountWithinBounds(principalAmount);
        this.assertInterestRateValid(interestRate);
        this.assertValidAmortizationUnit(amortizationUnit);
        this.assertTermLengthWholeAndWithinBounds(termLength);
        var interestRateFixedPoint = interestRate.mul(FIXED_POINT_SCALING_FACTOR);
        var principalTokenIndexHex = principalTokenIndex.toString(16);
        var principalAmountHex = principalAmount.toString(16);
        var interestRateFixedPointHex = interestRateFixedPoint.toString(16);
        var amortizationUnitTypeHex = AmortizationUnitCode[amortizationUnit.toUpperCase()].toString(16);
        var termLengthHex = termLength.toString(16);
        return ("0x" +
            principalTokenIndexHex.padStart(2, "0") +
            principalAmountHex.padStart(24, "0") +
            interestRateFixedPointHex.padStart(6, "0") +
            amortizationUnitTypeHex.padStart(1, "0") +
            termLengthHex.padStart(4, "0") +
            "0".repeat(27));
    };
    SimpleInterestLoanTerms.prototype.unpackParameters = function (termsContractParametersPacked) {
        this.assert.schema.bytes32("termsContractParametersPacked", termsContractParametersPacked);
        var principalTokenIndexHex = termsContractParametersPacked.substr(0, 4);
        var principalAmountHex = "0x" + termsContractParametersPacked.substr(4, 24);
        var interestRateFixedPointHex = "0x" + termsContractParametersPacked.substr(28, 6);
        var amortizationUnitTypeHex = "0x" + termsContractParametersPacked.substr(34, 1);
        var termLengthHex = "0x" + termsContractParametersPacked.substr(35, 4);
        var principalTokenIndex = new bignumber_1.BigNumber(principalTokenIndexHex);
        var principalAmount = new bignumber_1.BigNumber(principalAmountHex);
        var interestRateFixedPoint = new bignumber_1.BigNumber(interestRateFixedPointHex);
        var termLength = new bignumber_1.BigNumber(termLengthHex);
        // Given that our fixed point representation of the interest rate
        // is scaled up by our chosen scaling factor, we scale it down
        // for computations.
        var interestRate = interestRateFixedPoint.div(FIXED_POINT_SCALING_FACTOR);
        // Since the amortization unit type is stored in 1 byte, it can't exceed
        // a value of 255.  As such, we're not concerned about using BigNumber's
        // to represent amortization units.
        var unitCode = parseInt(amortizationUnitTypeHex, 16);
        // We only need to assert that the amortization unit type is valid,
        // given that it's impossible for the parsed totalExpectedRepayment,
        // principalTokenIndex, and termLength to exceed their bounds.
        this.assertValidAmortizationUnitCode(unitCode);
        var amortizationUnit = AmortizationUnitCode[unitCode].toLowerCase();
        return {
            principalTokenIndex: principalTokenIndex,
            principalAmount: principalAmount,
            interestRate: interestRate,
            termLength: termLength,
            amortizationUnit: amortizationUnit,
        };
    };
    SimpleInterestLoanTerms.prototype.assertPrincipalTokenIndexWithinBounds = function (principalTokenIndex) {
        if (principalTokenIndex.lt(0) || principalTokenIndex.gt(MAX_PRINCIPAL_TOKEN_INDEX_HEX)) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_TOKEN_INDEX(principalTokenIndex));
        }
    };
    SimpleInterestLoanTerms.prototype.assertPrincipalAmountWithinBounds = function (principalAmount) {
        if (principalAmount.lt(0) || principalAmount.gt(MAX_PRINCIPAL_AMOUNT_HEX)) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_PRINCIPAL_AMOUNT());
        }
    };
    SimpleInterestLoanTerms.prototype.assertInterestRateValid = function (interestRate) {
        var _a = interestRate.toString().split("."), rightOfDecimal = _a[1];
        var numDecimals = typeof rightOfDecimal !== "undefined" ? rightOfDecimal.length : 0;
        if (interestRate.lt(0) ||
            interestRate.gt(MAX_INTEREST_RATE) ||
            numDecimals > MAX_INTEREST_RATE_PRECISION) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_INTEREST_RATE());
        }
    };
    SimpleInterestLoanTerms.prototype.assertValidAmortizationUnitCode = function (amortizationUnitCode) {
        if (amortizationUnitCode > AmortizationUnitCodes.length - 1) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_AMORTIZATION_UNIT_TYPE());
        }
    };
    SimpleInterestLoanTerms.prototype.assertValidAmortizationUnit = function (amortizationUnitType) {
        if (!AmortizationUnitCodes.includes(amortizationUnitType)) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_AMORTIZATION_UNIT_TYPE());
        }
    };
    SimpleInterestLoanTerms.prototype.assertTermLengthWholeAndWithinBounds = function (termLengthInAmortizationUnits) {
        if (termLengthInAmortizationUnits.lt(0) ||
            termLengthInAmortizationUnits.gt(MAX_TERM_LENGTH_VALUE_HEX)) {
            throw new Error(exports.SimpleInterestAdapterErrors.INVALID_TERM_LENGTH());
        }
        else {
            this.assert.schema.wholeNumber("termLength", termLengthInAmortizationUnits);
        }
    };
    return SimpleInterestLoanTerms;
}());
exports.SimpleInterestLoanTerms = SimpleInterestLoanTerms;
var SimpleInterestLoanAdapter = /** @class */ (function () {
    function SimpleInterestLoanAdapter(web3, contracts) {
        this.assert = new invariants_1.Assertions(web3, contracts);
        this.contracts = contracts;
        this.termsContractInterface = new SimpleInterestLoanTerms(web3, contracts);
    }
    /**
     * Asynchronously generates a Dharma debt order given an instance of a
     * simple interest loan order.
     *
     * @param  simpleInterestLoanOrder a simple interest loan order instance.
     * @return                         the generated Dharma debt order.
     */
    SimpleInterestLoanAdapter.prototype.toDebtOrder = function (simpleInterestLoanOrder) {
        return __awaiter(this, void 0, void 0, function () {
            var principalTokenSymbol, principalAmount, interestRate, amortizationUnit, termLength, principalToken, principalTokenIndex, simpleInterestTermsContract, debtOrder;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.assert.schema.simpleInterestLoanOrder("simpleInterestLoanOrder", simpleInterestLoanOrder);
                        principalTokenSymbol = simpleInterestLoanOrder.principalTokenSymbol, principalAmount = simpleInterestLoanOrder.principalAmount, interestRate = simpleInterestLoanOrder.interestRate, amortizationUnit = simpleInterestLoanOrder.amortizationUnit, termLength = simpleInterestLoanOrder.termLength;
                        return [4 /*yield*/, this.contracts.loadTokenBySymbolAsync(principalTokenSymbol)];
                    case 1:
                        principalToken = _a.sent();
                        return [4 /*yield*/, this.contracts.getTokenIndexBySymbolAsync(principalTokenSymbol)];
                    case 2:
                        principalTokenIndex = _a.sent();
                        return [4 /*yield*/, this.contracts.loadSimpleInterestTermsContract(TX_DEFAULTS)];
                    case 3:
                        simpleInterestTermsContract = _a.sent();
                        debtOrder = omit(simpleInterestLoanOrder, [
                            "principalTokenSymbol",
                            "interestRate",
                            "amortizationUnit",
                            "termLength",
                        ]);
                        debtOrder = __assign({}, debtOrder, { principalToken: principalToken.address, termsContract: simpleInterestTermsContract.address, termsContractParameters: this.termsContractInterface.packParameters({
                                principalTokenIndex: principalTokenIndex,
                                principalAmount: principalAmount,
                                interestRate: interestRate,
                                amortizationUnit: amortizationUnit,
                                termLength: termLength,
                            }) });
                        return [2 /*return*/, types_1.DebtOrder.applyNetworkDefaults(debtOrder, this.contracts)];
                }
            });
        });
    };
    /**
     * Asynchronously generates a simple interest loan order given a Dharma
     * debt order instance.
     *
     * @param  debtOrder a Dharma debt order instance.
     * @return           the generated simple interest loan order.
     */
    SimpleInterestLoanAdapter.prototype.fromDebtOrder = function (debtOrder) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, principalTokenIndex, principalAmount, interestRate, termLength, amortizationUnit, principalTokenSymbol;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.assert.schema.debtOrderWithTermsSpecified("debtOrder", debtOrder);
                        _a = this.termsContractInterface.unpackParameters(debtOrder.termsContractParameters), principalTokenIndex = _a.principalTokenIndex, principalAmount = _a.principalAmount, interestRate = _a.interestRate, termLength = _a.termLength, amortizationUnit = _a.amortizationUnit;
                        return [4 /*yield*/, this.contracts.getTokenSymbolByIndexAsync(principalTokenIndex)];
                    case 1:
                        principalTokenSymbol = _b.sent();
                        return [4 /*yield*/, this.assertPrincipalTokenCorrespondsToSymbol(debtOrder.principalToken, principalTokenSymbol)];
                    case 2:
                        _b.sent();
                        return [2 /*return*/, __assign({}, debtOrder, { principalAmount: principalAmount,
                                principalTokenSymbol: principalTokenSymbol,
                                interestRate: interestRate,
                                termLength: termLength,
                                amortizationUnit: amortizationUnit })];
                }
            });
        });
    };
    /**
     * Asynchronously translates a Dharma debt registry entry into a
     * simple interest loan order.
     *
     * @param entry a Dharma debt registry entry
     * @return      the translated simple interest loan order
     */
    SimpleInterestLoanAdapter.prototype.fromDebtRegistryEntry = function (entry) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, principalTokenIndex, principalAmount, interestRate, termLength, amortizationUnit, principalTokenSymbol, loanOrder;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.assertIsSimpleInterestTermsContract(entry.termsContract)];
                    case 1:
                        _b.sent();
                        _a = this.termsContractInterface.unpackParameters(entry.termsContractParameters), principalTokenIndex = _a.principalTokenIndex, principalAmount = _a.principalAmount, interestRate = _a.interestRate, termLength = _a.termLength, amortizationUnit = _a.amortizationUnit;
                        return [4 /*yield*/, this.contracts.getTokenSymbolByIndexAsync(principalTokenIndex)];
                    case 2:
                        principalTokenSymbol = _b.sent();
                        loanOrder = {
                            principalTokenSymbol: principalTokenSymbol,
                            principalAmount: principalAmount,
                            interestRate: interestRate,
                            termLength: termLength,
                            amortizationUnit: amortizationUnit,
                        };
                        return [2 /*return*/, loanOrder];
                }
            });
        });
    };
    SimpleInterestLoanAdapter.prototype.getRepaymentSchedule = function (debtEntry) {
        var termsContractParameters = debtEntry.termsContractParameters, issuanceBlockTimestamp = debtEntry.issuanceBlockTimestamp;
        var _a = this.termsContractInterface.unpackParameters(termsContractParameters), termLength = _a.termLength, amortizationUnit = _a.amortizationUnit;
        return new types_1.RepaymentSchedule(amortizationUnit, termLength, issuanceBlockTimestamp.toNumber()).toArray();
    };
    SimpleInterestLoanAdapter.prototype.assertPrincipalTokenCorrespondsToSymbol = function (principalToken, symbol) {
        return __awaiter(this, void 0, void 0, function () {
            var addressMappedToSymbol;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contracts.getTokenAddressBySymbolAsync(symbol)];
                    case 1:
                        addressMappedToSymbol = _a.sent();
                        // Validate that the
                        if (principalToken !== addressMappedToSymbol) {
                            throw new Error(exports.SimpleInterestAdapterErrors.MISMATCHED_TOKEN_SYMBOL(principalToken, symbol));
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    SimpleInterestLoanAdapter.prototype.assertIsSimpleInterestTermsContract = function (termsContractAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var simpleInterestTermsContract;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contracts.loadSimpleInterestTermsContract()];
                    case 1:
                        simpleInterestTermsContract = _a.sent();
                        if (termsContractAddress !== simpleInterestTermsContract.address) {
                            throw new Error(exports.SimpleInterestAdapterErrors.MISMATCHED_TERMS_CONTRACT(termsContractAddress));
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    SimpleInterestLoanAdapter.Installments = {
        HOURLY: "hours",
        DAILY: "days",
        WEEKLY: "weeks",
        MONTHLY: "months",
        YEARLY: "years",
    };
    return SimpleInterestLoanAdapter;
}());
exports.SimpleInterestLoanAdapter = SimpleInterestLoanAdapter;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9;
