// @dharmaprotocol/contracts/dist/artifacts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const DebtRegistry_1 = require("./ts/DebtRegistry");
exports.DebtRegistry = DebtRegistry_1.DebtRegistry;
const DebtKernel_1 = require("./ts/DebtKernel");
exports.DebtKernel = DebtKernel_1.DebtKernel;
const RepaymentRouter_1 = require("./ts/RepaymentRouter");
exports.RepaymentRouter = RepaymentRouter_1.RepaymentRouter;
const TokenTransferProxy_1 = require("./ts/TokenTransferProxy");
exports.TokenTransferProxy = TokenTransferProxy_1.TokenTransferProxy;
const TokenRegistry_1 = require("./ts/TokenRegistry");
exports.TokenRegistry = TokenRegistry_1.TokenRegistry;
const DebtToken_1 = require("./ts/DebtToken");
exports.DebtToken = DebtToken_1.DebtToken;
const DummyToken_1 = require("./ts/DummyToken");
exports.DummyToken = DummyToken_1.DummyToken;
const ERC20_1 = require("./ts/ERC20");
exports.ERC20 = ERC20_1.ERC20;
const SimpleInterestTermsContract_1 = require("./ts/SimpleInterestTermsContract");
exports.SimpleInterestTermsContract = SimpleInterestTermsContract_1.SimpleInterestTermsContract;
const CollateralizedSimpleInterestTermsContract_1 = require("./ts/CollateralizedSimpleInterestTermsContract");
exports.CollateralizedSimpleInterestTermsContract = CollateralizedSimpleInterestTermsContract_1.CollateralizedSimpleInterestTermsContract;
const TermsContract_1 = require("./ts/TermsContract");
exports.TermsContract = TermsContract_1.TermsContract;
