// @dharmaprotocol/dharma.js/dist/lib/src/types/debt_order.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
// external
var bignumber_1 = require("../../utils/bignumber");
var moment = require("moment");
var assignDefaults = require("lodash.defaults");
// utils
var constants_1 = require("../../utils/constants");
var DebtOrder;
(function (DebtOrder) {
    DebtOrder.DEFAULTS = {
        kernelVersion: constants_1.NULL_ADDRESS,
        issuanceVersion: constants_1.NULL_ADDRESS,
        principalAmount: new bignumber_1.BigNumber(0),
        principalToken: constants_1.NULL_ADDRESS,
        debtor: constants_1.NULL_ADDRESS,
        debtorFee: new bignumber_1.BigNumber(0),
        creditor: constants_1.NULL_ADDRESS,
        creditorFee: new bignumber_1.BigNumber(0),
        relayer: constants_1.NULL_ADDRESS,
        relayerFee: new bignumber_1.BigNumber(0),
        underwriter: constants_1.NULL_ADDRESS,
        underwriterFee: new bignumber_1.BigNumber(0),
        underwriterRiskRating: new bignumber_1.BigNumber(0),
        termsContract: constants_1.NULL_ADDRESS,
        termsContractParameters: constants_1.NULL_BYTES32,
        expirationTimestampInSec: new bignumber_1.BigNumber(moment()
            .add(30, "days")
            .unix()),
        salt: new bignumber_1.BigNumber(0),
        debtorSignature: constants_1.NULL_ECDSA_SIGNATURE,
        creditorSignature: constants_1.NULL_ECDSA_SIGNATURE,
        underwriterSignature: constants_1.NULL_ECDSA_SIGNATURE,
    };
    function applyNetworkDefaults(debtOrder, contracts) {
        return __awaiter(this, void 0, void 0, function () {
            var debtKernel, repaymentRouter, networkDefaults;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, contracts.loadDebtKernelAsync()];
                    case 1:
                        debtKernel = _a.sent();
                        return [4 /*yield*/, contracts.loadRepaymentRouterAsync()];
                    case 2:
                        repaymentRouter = _a.sent();
                        networkDefaults = __assign({}, DebtOrder.DEFAULTS, { kernelVersion: debtKernel.address, issuanceVersion: repaymentRouter.address });
                        return [2 /*return*/, assignDefaults(debtOrder, networkDefaults)];
                }
            });
        });
    }
    DebtOrder.applyNetworkDefaults = applyNetworkDefaults;
})(DebtOrder = exports.DebtOrder || (exports.DebtOrder = {}));
